etl/command.py

```diff
@@ -357,16 +376,53 @@ def run_dag(
             f"--- Would run {len(steps)} steps{_create_expected_time_message(total_expected_time_seconds, prepend_message=' (at least ')}:"
         )
         return enumerate_steps(steps)
-    elif workers == 1:
-        print(
-            f"--- Running {len(steps)} steps{_create_expected_time_message(total_expected_time_seconds, prepend_message=' (at least ')}:"
-        )
-        return exec_steps(steps, strict=strict)
-    else:
-        print(
-            f"--- Running {len(steps)} steps with {workers} processes ({config.GRAPHER_INSERT_WORKERS} threads each):"
-        )
-        return exec_steps_parallel(steps, workers, dag=dag, strict=strict)
+
+    if engine == "etl":
+        if workers == 1:
+            print(
+                f"--- Running {len(steps)} steps{_create_expected_time_message(total_expected_time_seconds, prepend_message=' (at least ')}:"
+            )
+            return exec_steps(steps, strict=strict)
+        else:
+            print(
+                f"--- Running {len(steps)} steps with {workers} processes ({config.GRAPHER_INSERT_WORKERS} threads each):"
+            )
+            return exec_steps_parallel(steps, workers, dag=dag, strict=strict)
+
+    elif engine == "prefect":
+        if workers == 1:
+            # task_runner = task_runners.SequentialTaskRunner()
+            task_runner = prefect.task_runners.ConcurrentTaskRunner()
+        else:
+            task_runner = DaskTaskRunner(cluster_kwargs={"n_workers": workers, "threads_per_worker": 1})
+
+        def on_failure_hook(task: prefect.Task, task_run: TaskRun, state: prefect.State):
+            """Cancel all tasks if one fails."""
+            with get_dask_client() as client:
+                print("Failure detected. Shutting down Dask client and cancelling all tasks.")
+                # force kill Dask client
+                client.shutdown()
+
+        @prefect.flow(log_prints=True, task_runner=task_runner)
+        def run_etl(steps):
+            task_futures: Dict[str, PrefectFuture] = {}
+
+            for step in steps:
+                # task = prefect.task(name=str(step), on_failure=[on_failure_hook])
```

Comment:
I am noticing that the flows hang around forever on failure. Is there a reason this bit is commented out?

Reply:
Hmm, when does it happen? When I run it locally with an error, it doesn't hang **forever**, but it fails after all tasks are either completed or failed.